packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -910,12 +930,69 @@ class PerformanceOverlayLayer extends Layer {
   }
 }
 
+/// The signature of the callback added in [ContainerLayer.addCompositionCallback].
+typedef CompositionCallback = void Function(ContainerLayer);
+
 /// A composited layer that has a list of children.
 ///
 /// A [ContainerLayer] instance merely takes a list of children and inserts them
 /// into the composited rendering in order. There are subclasses of
 /// [ContainerLayer] which apply more elaborate effects in the process.
 class ContainerLayer extends Layer {
+  static int _nextCallbackId = 0;
+
+  @override
+  void _fireCompositionCallbacks(bool includeChildren) {
+    super._fireCompositionCallbacks(includeChildren);
+    if (!includeChildren) {
+      return;
+    }
+    Layer? child = firstChild;
+    while (child != null) {
+      child._fireCompositionCallbacks(includeChildren);
+      child = child.nextSibling;
+    }
+
+  }
+
+  /// Adds a callback for when the layer tree that this layer is part of gets
```

Comment:
Conceptually, why is this a feature of ContainerLayer over the Layer base class?

Reply:
Mainly because RenderObjects are the normal clients for this via the PaintingContext, and the PaintingContext only _knows_ its container layer - picture layers might not ever be created for it.

Reply:
Since the backing storage is in the Layer class, it feels like this method should also be defined there? PaintingContext could still use it on the ContainerLayer, right?

Reply:
That will then require casts to get to parents, but I guess that's ok.

Reply:
oh, it won't. I see now.

Reply:
It will but it's probably ok.